mautic/mautic
# [DPMMA-2075] Including totals row in report export and increased precision for average

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [:x:]
| New feature/enhancement? (use the a.x branch)      | [:white_check_mark:]
| Deprecations?                          | [:x:]
| BC breaks? (use the c.x branch)        | [✅]
| Automated tests included?              | [✅] 
| Related user documentation PR URL      | #
| Related developer documentation PR URL | #
| Issue(s) addressed                     | #

#### Description:

This PR adds a Totals row in report CSV/HTML/Excel exports - [Mautic Forum: totals-in-report-rounding-to-4-decimal-places-and-including-in-export](https://forum.mautic.org/t/totals-in-report-rounding-to-4-decimal-places-and-including-in-export/26479). Calculating totals has been moved from report view template to ReportDataResult class. PR also sets the precision of total average to 4 decimal places.

#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create report with standard and calculated columns e.g. 
![report_1](https://user-images.githubusercontent.com/102536220/229114935-16241ff5-afea-43d5-86ba-211ac0f055b7.jpg)
3. In report view (s/reports/view/{id}/{page}), in the Totals row, the calculated average should be accurate to 4 decimal places. The other values will be valid but not changed from the user's point of view.
![report_2](https://user-images.githubusercontent.com/102536220/229117144-00938287-9fe4-477b-9fa6-5b46e7f53de3.jpg)
4. In the HTML export view (s/reports/view/{id}/export/html) Totals will only be added if the report contains at least one calculated column.
![report_3](https://user-images.githubusercontent.com/102536220/229117344-4d2914a0-e957-4bac-8aba-646ca6d51266.jpg)
6. In the CSV/Excel export, the Totals row will also be included only if the report contains at least one calculated column. "Totals" label is translated and placed in the leftmost cell (if is not calculated column and cell is empty) because there is no additional ordinal column with row numbers in these exports as in the HTML preview or report.
![report_4](https://user-images.githubusercontent.com/102536220/229118448-c3a96b96-6f0c-4a5c-a51d-6cde3e820691.jpg)
![report_5](https://user-images.githubusercontent.com/102536220/229118404-c7a19041-1994-4403-b9ad-ad2e0048bcef.jpg)
